### DISCUSSION TITLE
Whisper iOS transcription app

### DISCUSSION THREAD

**Discussion by bjnortier** [openai/whisper#443] [Show and tell]
This is based on the whisper.cpp implementation by @ggerganov. It uses the tiny model and all processing is done on-device.

It's in the App Store:
https://apps.apple.com/za/app/hello-transcribe/id6443919768

**Comment by ggerganov** [comment]
Awesome!
There is an idea to shrink the model using [8-bit weights + scaling factors] instead of 16-bit weights, which would reduce the app size by factor of 2. But the accuracy might suffer too much.

**Comment by nyadla-sys** [comment]
@bjnortier is it possible to implement the iOS app based on [whisper.tflite](https://github.com/usefulsensors/openai-whisper/blob/main/models/whisper.tflite)(40MB model weights are in int8)

Refer this to run inference in python
https://colab.research.google.com/github/usefulsensors/openai-whisper/blob/main/notebooks/tflite_from_huggingface_whisper.ipynb

Refer README to run inference on C++
https://github.com/usefulsensors/openai-whisper

to run tflite files on iOS refer the below link 
https://www.tensorflow.org/lite/guide/ios

**Reply by bjnortier** [reply]
Thanks, I was looking at this last night. Seems slower than @ggerganov's C++ on my M1 Max. Can it use the CoreML delegate on iOS? Curious about performance differences

**Reply by bjnortier** [reply]
Seems like I would need tom implement a custom Tensorflow Task?

**Reply by nyadla-sys** [reply]
As per below it accelerates tflite on iOS.
On iOS devices, you can enable use of GPU-accelerated execution of your models using a delegate. Delegates act as hardware drivers for TensorFlow Lite, allowing you to run the code of your model on GPU processors. This page describes how to enable GPU acceleration for TensorFlow Lite models in iOS apps.

https://www.tensorflow.org/lite/ios/delegates/gpu
https://www.tensorflow.org/lite/performance/coreml_delegate

**Reply by nyadla-sys** [reply]
> Seems like I would need tom implement a custom Tensorflow Task?


> Seems like I would need tom implement a custom Tensorflow Task?

Could you please implement an iOS app using whisper.tflite model ?

**Reply by bjnortier** [reply]
I'm looking into it... I had some issues getting the TFLite Sound Classifier example app to work, but it seems doable using the C++ log Mel spectrogram

**Reply by nyadla-sys** [reply]
Please use c++ logmel spectrogram and use inference using tflite framework with coreml/gpu delegate 
it may perform better as most of the kernels optimized for iOS using coreml/gpu delegate

**Reply by impara** [reply]
Other than speed, what is the benefit or the differences between the .tflite & the .cpp implementations?

**Reply by bjnortier** [reply]
.cpp version has a but more functionality, and I could get .tflite to work at all with the bigger multilingual models

**Comment by Abdurrafey-Siddiqui** [comment]
> This is based on the whisper.cpp implementation by @ggerganov. It uses the tiny model and all processing is done on-device.
> 
> It's in the App Store: https://apps.apple.com/za/app/hello-transcribe/id6443919768

Hi could you also add the options to tweak the transcribe settings for example temperature, initial_prompt ect...

**Reply by bjnortier** [reply]
Those options are in the original OpenAI python implementation but are not available in the .cpp implementation I'm using

**Comment by chrismarquardt** [comment]
This is pretty awesome. Thanks for making this!. I assume you're using the English-only tiny.en model? Just wondering for us multi-linguals. The models perform quite well in some other languages too.

**Comment by floriandotorg** [comment]
Awesome project! Was looking for something exactly like this. I think multi-language support, as well as at least the medium model would be necessary to make this really useful.

Also, if you'd add in-app audio recording and live-transcribtion, it would easily be the best transcribing app in the App Store.

**Reply by bjnortier** [reply]
Working on it!

**Comment by AIceegg** [comment]
congrats to this amazing project, @bjnortier! this is so much better than current mobile transcription apps! are you planning to implement multi-language (f.e. german) and in-app features? would love to support you with that - unfortunately, not my area.

**Comment by v-puppala** [comment]
@bjnortier this is really cool! Which language did you use? Objective C or Swift? Also was there any preprocessing you did with the raw audio data collected by the app prior to feeding it into the model?

**Reply by bjnortier** [reply]
Swift & SwiftUI, but it uses the C++ implementation of Whisper.
The only preprocessing currently is to convert to 16-bit 16kHz PCM (WAV) from other formats.

**Comment by TerenzeY** [comment]
Love this, but mine seems to be crashing whenever I tap on the Transcribe button. I must be doing something wrong..?

```
Initializing Whisper...
whisper_model_load: invalid model file '/private/var/containers/Bundle/Application/87E677AB-2598-4D1F-88A5-DE72C1E1DDF9/whisper-ios-demo.app/ggml-tiny.bin' (bad magic)
whisper_init: failed to load model from '/private/var/containers/Bundle/Application/87E677AB-2598-4D1F-88A5-DE72C1E1DDF9/whisper-ios-demo.app/ggml-tiny.bin'
opened WAV. size: 992576
```

**Reply by bjnortier** [reply]
Seems like the ggml-tiny.bin is not correct. What's the file size?

**Reply by TerenzeY** [reply]
133 bytes...

**Reply by TerenzeY** [reply]
How can I compile the right one?

**Reply by TerenzeY** [reply]
any thoughts @bjnortier ?

**Reply by bjnortier** [reply]
A. You are asking questions about the demo app I put on Github, not the app
I have on the App Store mentioned in this thread, so pls ask question there.
B. You need to install Git LFS to get the correct binary file, it’s 74.1MB as you can see here:
https://github.com/bjnortier/whisper-ios-demo/blob/main/resources/ggml-tiny.bin

**Reply by TerenzeY** [reply]
got it thanks.

**Comment by jsharf** [comment]
This needs apple watch integration! It would be awesome if I could just talk into my watch from time to time and get a transcription of it on my phone later.

**Reply by bjnortier** [reply]
I'll have to get a watch then :)

**Reply by whitesided** [reply]
Seems like jsharf is volunteering to get you one for the holiday season!

**Comment by theos2node** [comment]
Hey, I love what you're doing here. I'm kind of new to coding I'm not too sure how this works. But I'm curious if this app takes advantage of Apple silicon?

I know this must be a lot to do, but a really cool suggestion: able to attach the audio files with the transcription and to be able to scrub through the audio file using the transcription. Also, a search bar so you can search your personal library of transcriptions.

Example: say I'm looking for the word "Jeremy," and so I search it up, and it will highlight all the moments in the audio file where it said Jeremy, and you can click it, and it will bring you exactly where it was said on the audio file (you can even have a broad search for all files!!!)

These are only suggestions:))
I love what you're doing!!!!!

**Reply by bjnortier** [reply]
Awesome suggestions, thanks. There's an app that has some of these features already if you want to try it: https://apps.apple.com/za/app/transcribe-speech-to-text/id1241342461

In the long run I would be able to add this but I have to see if it's commercially viable to put the effort in

**Comment by whitesided** [comment]
Any chance you could update the platforms so this could be run on an Apple silicon Mac? I run the original python via pipx but I know a few people who'd be delighted to fiddle with it.

**Reply by bjnortier** [reply]
Possible, just not a high priority right now, will add it to the list :)

**Reply by sawhney17** [reply]
Check out the MacWhsiper app!

**Comment by UEUEE** [comment]
Hi, I have a question We need to recognize a specific phrase. And make recognized words bold. Can we use whisper? 
I have checked your app and found it takes pretty much time before words appeared. Does it wait till the end of the phrase before show it?

**Reply by bjnortier** [reply]
Do you mean like "Hey Siri"? It's a very specific phrase and I don't think you need such big (i.e. slower) model to detect a specific phrase. My app does dictation pretty much as fast as the processor can handle it. I.e. it does a transcription, while that's happening it accumulates more audio samples, when it's done it runs again on the current samples.

**Comment by cyturralde** [comment]
This is a great app. Just curious. Is there a way to make it differentiate between speakers?

**Reply by bjnortier** [reply]
Thanks! The answer is no currently but could be added, just not sure what you mean by differentiate?

**Reply by cyturralde** [reply]
In this case I was thinking it would distinguish between speakers and transcribe it out more like a chat like this:

Person 1: text that person one spoke
Person 2: text the person two spoke
Person 1: some more of person one speaking
Person 2: more of person two speaking.

and so on and so on. I think this would be really useful for basically all meetings of various kinds.

**Reply by bjnortier** [reply]
Ah I see. Thanks. Yes very useful for meetings. 
I don't think Whisper has that built-in, so you would have to use another model.
Some discussion here:
https://github.com/openai/whisper/discussions/264

But yes would be very useful. Will add it to the list of future features

**Comment by sawhney17** [comment]
Differentiation in this case would refer to speaker diarization. i.e telling which speaker is speaking what and like splitting the transcription

**Reply by bjnortier** [reply]
Cool, thanks

**Comment by azertiv** [comment]
Thank you for this very practical application! Several suggestions:
- Add a widget on the lock screen, which would allow you to save an idea directly and quickly;
- Be able to dictate text directly from the keyboard, but I guess it's complex to implement...

**Reply by bjnortier** [reply]
1. Hi, the next version will include Shortcuts, so you will be able to create a shortcut to create a note etc., which you can use via Siri. A widget would be useful... will add it to the feature requests
2. I've investigate the keyboard idea and discussed with someone who also wanted that feature. iOS has a hard memory limit for keyboards (device dependent, on mine it's 70MB I think), but it's too small to load even the "tiny" model. This is to ensure a keyboard is fast. The alternative is to make a keyboard that uses a server-side transcription engine, which has privacy implications. I am considering adding server-side transcription for people who do want to use it.

**Reply by antgiant** [reply]
I don't know if it helps but a while ago I noticed that Microsoft's SwiftKey Keyboard launched an app when you pressed the transcribe button. (Looks like the don't anymore though.) Makes me wonder if they were working around some limitation by launching the full app and starting a local server so that they could stream the audio that way.

**Reply by dvkch** [reply]
A shortcut would be awesome ! I'd like to be able to make a shortcut to run a video through this app, transcribe once in original language, transcribe a second time in english, and add both subtitles to the video, to be able to send videos to my barely-speaking french friends that want to learn a bit 😊 that would allow them to see everyday phone videos with both french and english subtitles !

If you ever implement a shortcut, it would be amazing to select the language and the english translation when using it.

**Reply by aehlke** [reply]
@bjnortier how about app extensions? or app clips? not sure if either are usable from within keyboard

**Reply by bjnortier** [reply]
@aehlke I haven't tried but I think it won't be possible due to memory constraints.

**Comment by RootMePLS** [comment]
Application is crashed after upgrading to PRO(By all Pro features) and turning on CoreML.
```
iPhone 11
iOS 16.4.1
App version 2.3.0
```

After reinstalling the app, I tried to "Restore Purchases", but nothing happened, only after uploading and restarting the app, download buttons appears.

Steps to reproduce:
- Upgrade to PRO
- Download Accurate Multi-lang
- Turn on "Use CoreML"
- Choose Accurate Multi-lang
- Crash

My proposal is, to block Turn on "Use CoreML" if iPhone model < 13

**Reply by bjnortier** [reply]
Thanks, I will look into why Restore doesn't work :/

I don't want to block CoreML because some people have been able to get it working on an iPhone 12, but I could it off by default for < 13, which would help. That is why I have added messaging and warnings about CoreML to 2.3.0, so people can decide for themselves

**Reply by RootMePLS** [reply]
Got it

**Comment by chrismarquardt** [comment]
Loving on-device quality transcription, thanks for making this! I now tried to automate a few things, what I was, hoping was to be able to seamlessly integrate whisper into a shortcut, and then use its output text in the next step. but the shortcuts don’t seem to work on iOS 16.5

“Invalid action metadata
Unfortunately, there is nothing
Shortcuts can do about this; it is due to a bug in the host app. We'd recommend contacting the maker of that app and letting them know of this problem.”

![IMG_3180](https://github.com/openai/whisper/assets/2527608/d9d271c6-81b8-4d3c-84da-e3ab0fcf8f5a)

**Reply by bjnortier** [reply]
Hi, ok thanks for letting me know. I'm updating mine to 16.5 now to diagnose the problem

**Reply by bjnortier** [reply]
Hi Chris, I've just updated to 16.5 and my shortcuts seem to work (both the dictate and File shortcuts). Can you please tell me how you produced this error? Secondly, what version of Hello Transcribe are you using? 2.4 went live yesterday so can you try to update if you haven't already? (don't think that'll make a difference, but at least we'll work from the same starting point).
Thanks!

**Reply by chenzhony** [reply]
Can you provide the shortcut file you are currently using? 
I am not familiar with shortcuts, so I don't know how to use them.
At the same time, I was thinking that it would be great if I could directly modify Apple's built-in dictation to yours, so that it can be used directly in all apps. Or is there possible to make this into an input method, which also can be used directly in all apps

**Reply by bjnortier** [reply]
This is the shortcut to create a Note from a file a user selects:
https://www.icloud.com/shortcuts/f9c9d2ef109346bd997e6b64c7c184ae

You can't modify Apple's built-in dictation engine, and it's not possible to build a Whisper keyboard because keyboards have very strict memory limitations

**Reply by bjnortier** [reply]
@chrismarquardt sorry I didn't see the image you posted. Tried to re-create but don't see that error :/

![IMG_5816](https://github.com/openai/whisper/assets/57994/4f41e96d-d173-49ef-af8c-2c1422b41d02)

**Reply by chrismarquardt** [reply]
After rebooting the phone, it has worked fine a couple of times, but I see intermittent errors after hitting the stop button in Hello Transcribe to finish the dictation and continue the shortcut.

Here are some simple test shortcuts that intermittently fail for me:

1. dictation test: https://www.icloud.com/shortcuts/571801df774f4bfeb25740dc53f49ad6

2. transcription test: https://www.icloud.com/shortcuts/571801df774f4bfeb25740dc53f49ad6

Some of the failures seem due to HT terminating while running, some feel like something under shortcuts' hood gets stuck.

The most reliable way so far seems to use the built-in recorder to record to a file, then transcribe that file using Hello Transcribe as shown in this shortcut: https://www.icloud.com/shortcuts/215182e1c9dc4cd3a38f018addffc8a5

Here are a couple of the intermittent error messages:
![IMG_3194](https://github.com/openai/whisper/assets/2527608/855744f9-8e71-472b-9e9a-796499b8278a)
![IMG_3193](https://github.com/openai/whisper/assets/2527608/5e8223c2-7ab0-4627-ad7c-593c5e0bc3d6)

**Reply by bjnortier** [reply]
Thanks. There is an issue with a CoreML memory leak. Trying to get a fix for that out asap then will look into this

**Reply by bjnortier** [reply]
Can I add you to TestFlight? Then you can check the latest Beta. There have been some issues with Dictation

**Reply by chrismarquardt** [reply]
Sure thing, I'd love to give it a spin and report back.

**Reply by ninopiamonte** [reply]
Hi! Can I try the beta too? I've been running into the same timeout problems with longer sessions. Thank you

**Reply by bjnortier** [reply]
@ninopiamonte Added you 👍

**Reply by ninopiamonte** [reply]
Thank you! I've been testing since you added me and the "small.en" model works fine on my iPhone 13.

**Comment by chrismarquardt** [comment]
By the way, kudos for getting CoreML to work. I run this on an iPhone 13 pro and it's orders of magnitude faster than just on CPU. How is this even possible?

**Reply by bjnortier** [reply]
Can't really take credit for the CoreML work, I can only take credit for integrating it. It's pretty impressive.

**Comment by josegomez** [comment]
can you add the option to translate to other languages than english?

**Reply by bjnortier** [reply]
Whisper doesn't have support for other languages, if I were to add it it would be a separate system (i.e. major effort).

**Reply by josegomez** [reply]
ah , got it I thought the translation was using another service already.

**Comment by 7k50** [comment]
Just installed Hello Transcribe for iPhone X. Seems not to work, can't get any results: I only see a blue bar repeatedly going from left to right. Perhaps this app isn't compatible with iPhone X? If so, it might be possible to disable download for this app for unsupported devices in apples App Store.

**Reply by bjnortier** [reply]
It's probably just really slow on the iPhone X. I have users on iPhone 8 who use it it's just really slow so I don't want to disable devices.

**Reply by 7k50** [reply]
> It's probably just really slow on the iPhone X. I have users on iPhone 8 who use it it's just really slow so I don't want to disable devices.

You're right, it worked once I waited a few more moments. In that case, please keep it enabled for iPhone X and under. The speed is still relatively useful. I was probably tricked by the constantly resetting blue progress bar. I think that this UX, when it's constantly resetting like that, gives the user (me) the intuition that the app is broken, when also coupled with a slower than average transcription speed as in my case.

**Reply by bjnortier** [reply]
So the progress bar has a transparent indicator and a solid indicator, I tried to make a UI that doesn't need another spinner or something, but I can see it's confusing. I'll change it

**Comment by chrismarquardt** [comment]
Feature request for the Shortcuts integration: Would it be possible for the multilinguals among us to make a language selection when calling the Transcribe action from Shortcuts? The reason is the higher CoreML speed and efficiency with a language pre-selected.

**Reply by bjnortier** [reply]
Ah yes good idea, will add it to the backlog

**Comment by TFWol** [comment]
What is http://hello-transcribe.b-cdn.net/ ? The app beaconed to that url.

**Reply by TFWol** [reply]
@bjnortier bump

**Reply by bjnortier** [reply]
Sorry didn't see this. It's the CDN (bunny.net) for the Whisper models.

**Reply by TFWol** [reply]
Gotcha, thanks!

**Comment by Aamirfarooq13** [comment]
I'm looking app or shortcut for transcription multilingual OpenAi whisper v3 support for iPhone 14 Pro Max, anyone know about this ?

**Reply by bjnortier** [reply]
You can use my app, Large V3 is supported in the Beta, can add you to TestFlight if you'd like (send me your email address).

https://apple.co/49eTzRn

**Reply by Aamirfarooq13** [reply]
I'm using Hello Transcribe app in my iPhone 14 Pro Max with iOS 17.4.1
Thanks for one time purchase option into app very reasonable price, I bought now I'm tension free no more minutes limit and nothing else, my reason to use this app I run small YouTube channel for that I record my video in my native language Urdu (Pakistan) before upload to YouTube I need SRT file for because when I upload video with SRT file YouTube shows close caption to my videos, it is big benefit YouTube algorithm to understand my text translate in other languages automatically I get global audience reach, this is behind me whole idea, Hello Transcribe today I used after long time because when I was using whisper v3 crash app but today this not happen maybe you fixed issue, anyway, my big concerns I require fine tune model for Urdu (Pakistan) language, Hello Transcribe app voice recognition skip some word in Urdu language, if you can fix this issue it will more useful for me, I hope you find this my message

**Comment by mak448a** [comment]
Will iOS 15 support ever be added?

**Comment by ponmike92** [comment]
Thank you for such an amazing implementation!

is it possible to use large-v3-turbo?
Haven’t found it in the list
